transaction/std/src/signed_contingent_input_builder.rs

```diff
@@ -282,6 +294,173 @@ impl<FPR: FogPubkeyResolver> SignedContingentInputBuilder<FPR> {
+        let mut mb = self
+            .memo_builder
+            .take()
+            .expect("memo builder is missing, this is a logic error");
+        let result = self.make_fractional_output_with_fog_hint_address(
+            amount,
+            &change_destination.change_subaddress,
+            &change_destination.primary_address,
+            |memo_ctxt| mb.make_memo_for_change_output(amount, change_destination, memo_ctxt),
+            rng,
+        );
+        // Put the memo builder back, BEFORE question mark operator
+        self.memo_builder = Some(mb);
+        let (revealed_tx_out, confirmation_number) = result?;
+        self.fractional_change = Some(revealed_tx_out.clone());
+        Ok((revealed_tx_out.tx_out, confirmation_number))
+    }
+
+    /// Add a required output to the rules, using `fog_hint_address` to
+    /// construct the fog hint.
+    ///
+    /// This is a private implementation detail, and generally, fog users expect
+    /// that the transactions that they receive from fog belong to the account
+    /// that they are using. The only known use-case where recipient and
+    /// fog_hint_address are different is when sending change transactions
+    /// to oneself, when oneself is a fog user. Sending the change to the
+    /// main subaddress means that you don't have to hit fog once for the
+    /// main subaddress and once for the change subaddress, so it cuts the
+    /// number of requests in half.
+    ///
+    /// # Arguments
+    /// * `amount` - The amount of this output
+    /// * `recipient` - The recipient's public address
+    /// * `fog_hint_address` - The public address used to create the fog hint
+    /// * `memo_fn` - The memo function to use (see TxOut::new_with_memo)
+    /// * `rng` - RNG used to generate blinding for commitment
+    fn make_fractional_output_with_fog_hint_address<RNG: CryptoRng + RngCore>(
+        &mut self,
+        amount: Amount,
+        recipient: &PublicAddress,
+        fog_hint_address: &PublicAddress,
+        memo_fn: impl FnOnce(MemoContext) -> Result<MemoPayload, NewMemoError>,
+        rng: &mut RNG,
+    ) -> Result<(RevealedTxOut, TxOutConfirmationNumber), TxBuilderError> {
+        if !self.block_version.masked_amount_v2_is_supported() {
+            return Err(TxBuilderError::FeatureNotSupportedAtBlockVersion(
+                *self.block_version,
+                "partial fills",
+            ));
+        }
+
+        let (hint, pubkey_expiry) =
+            crate::transaction_builder::create_fog_hint(fog_hint_address, &self.fog_resolver, rng)?;
+
+        let (tx_out, shared_secret) = crate::transaction_builder::create_output_with_fog_hint(
+            self.block_version,
+            amount,
+            recipient,
+            hint,
+            memo_fn,
+            rng,
+        )?;
+        self.impose_tombstone_block_limit(pubkey_expiry);
+
+        let amount_shared_secret = MaskedAmountV2::compute_amount_shared_secret(&shared_secret);
```

Comment:
This is possibly an abstraction leakage, maybe it should be like

```
tx_out.masked_amount.compute_amount_shared_secret(...)
```

or something?